Add support for directories in url of install command

horike37 opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

Closes #2721

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
- Download zip that the Github repo whole things.
- Delete except for specific directories.
<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:
Please run as follow:
```
serverless install --url=https://github.com/serverless/examples/tree/master/aws-node-rest-api-with-dynamodb
```
<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* AWS CLI commands - To list AWS resources and show that the correct config is in place
* Other - Anything else that comes to mind to help us evaluate
-->


## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below


***Is this ready for review?:*** YES

mthenw commented on docs/providers/aws/cli-reference/install.md:
````diff
@@ -32,4 +32,12 @@ serverless install --url https://github.com/some/service
+### Installing a service from directories in a GitHub URL
+
+```bash
+serverless install --url https://github.com/johndoe/authentication/tree/master/todo/someauth
````
could you use https://github.com/serverless/examples/tree/master/rest-api-with-dynamodb in example?

mthenw commented on lib/plugins/install/install.js:
```diff
@@ -67,23 +67,40 @@ class Install {
-    if (this.serverless.utils.dirExistsSync(servicePath)) {
-      const errorMessage = `A folder named "${parsedGitHubUrl.repo}" already exists.`;
-      throw new this.serverless.classes.Error(errorMessage);
+    const endIndex = parts.length - 1;
```
this whole code block can be simplified

```javascript
    let dirName;
    if (parts.length > 4) {
      dirName = parts[endIndex];
    } else {
      dirName = parsedGitHubUrl.repo;
    }

   if (this.serverless.utils.dirExistsSync(servicePath)) {
      const errorMessage = `A folder named "${dirName}" already exists.`;
      throw new this.serverless.classes.Error(errorMessage);
    }
    this.serverless.cli.log(`Downloading and installing "${dirName}"...`);
```

mthenw commented on lib/plugins/install/install.js:
```diff
@@ -67,23 +67,40 @@ class Install {
+        for (let i = 5; i <= endIndex; i++) {
+          dirctory = path.join(dirctory, parts[i]);
+        }
+        that.serverless.utils.copyDirContentsSync(dirctory, repoPath);
```
`that.serverless.utils.copyDirContentsSync(dirctory, path.join(process.cwd(), parts[endIndex]));`

mthenw commented on lib/plugins/install/install.js:
```diff
@@ -67,23 +67,40 @@ class Install {
+        }
+        that.serverless.utils.copyDirContentsSync(dirctory, repoPath);
+        fs.removeSync(servicePath);
+        that.serverless.cli.log(`Successfully installed "${parts[endIndex]}".`);
```
tan be removed

mthenw commented on lib/plugins/install/install.js:
```diff
@@ -67,23 +67,40 @@ class Install {
+        fs.removeSync(servicePath);
+        that.serverless.cli.log(`Successfully installed "${parts[endIndex]}".`);
+      } else {
+        that.serverless.cli.log(`Successfully installed "${parsedGitHubUrl.repo}".`);
```
`else` can also be removed and 

```
 that.serverless.cli.log(`Successfully installed "${dirName}".`);
```

should be added at the end

horike37 commented:
@mthenw 
Thank you for reviewing! I fixed all. Please check it

pmuens commented on docs/providers/aws/cli-reference/install.md:
```diff
@@ -32,4 +32,12 @@ serverless install --url https://github.com/some/service
 
-This example will download the .zip file of the `authentication` service from GitHub, create a new directory with the name `authentication` in the current working directory and unzips the files in this directory.
\ No newline at end of file
+This example will download the .zip file of the `authentication` service from GitHub, create a new directory with the name `authentication` in the current working directory and unzips the files in this directory.
```
This text needs to be updated as well (so that it mentions that the DynamoDB service is used).

pmuens commented on docs/providers/aws/cli-reference/install.md:
````diff
@@ -32,4 +32,12 @@ serverless install --url https://github.com/some/service
+serverless install --url https://github.com/serverless/examples/tree/master/rest-api-with-dynamodb
+```
+
+This example will download the `rest-api-with-dynamodb` service from directory in GitHub.
````
`from a directory in GitHub`

horike37 commented:
@pmuens 
Thank you! I fixed the docs!

pmuens commented on lib/plugins/install/install.js:
```diff
@@ -67,23 +66,34 @@ class Install {
+      if (parts.length > 4) {
+        let dirctory = servicePath;
+        for (let i = 5; i <= endIndex; i++) {
+          dirctory = path.join(dirctory, parts[i]);
```
`directory`

horike37 commented:
@pmuens 
thx! Just fixed typo.

pmuens commented:
Thanks @horike37 💯 

Unfortunately I still get the `example` directory donwloaded alongside the services directory...

horike37 commented:
@pmuens @mthenw 
I thought it was better to not leave `examples` in the review.

I first implemented as follow:
- Download the whole a zip from the repository and expanded the zip.
- Moved target directory to top level directory.
- Left the necessary parts.(used by https://github.com/serverless/serverless/pull/2816/files/0e274d448de4a9327824d77f7e7802ea26e3a4c0#diff-8f56e8b98ccd6325e189af4336830e28R99)

But he said to be removed `fs.removeSync` in the review, then I deleted this implementation.
Can I remain this, then delete the whole thing like the `examples` directory?

mthenw commented:
@horike37 hmm, imho it should look like this:

- download whole zip to temp dir,
- extract
- copy only picked directory to the current dir
- delete temp dir

horike37 commented:
@mthenw
I got it! I will send PR again and fix it.
Thank you for the advice!

pmuens commented:
@horike37 thanks for jumping in again! 🎉

horike37 commented:
@pmuens @mthenw 
Fixed the flow of processing. It remained only picked directory.
Please check it.

thanks!
